c/ovm.c

```diff
@@ -490,39 +492,42 @@ static word prim_ref(word pword, word pos) {
 }
 
 static int64_t cnum(word a) {
+   uint64_t x;
    if (allocp(a)) {
-      int64_t x = 0;
-      int shift = 0;
       word *p = (word *)a;
-      while (a != INULL) {
-         x |= F(p[1]) << shift;
+      unsigned int shift = 0;
+      x = 0;
+      do {
+         x |= fixval(p[1]) << shift;
          shift += FBITS;
          p = (word *)p[2];
-      }
-      return x;
+      } while (allocp(p));
+      return header(a) == NUMNHDR ? -x : x;
    }
-   return fixval(a);
+   x = fixval(a);
+   return imm_type(a) == TNUMN ? -x : x;
 }
 
 static word onum(int64_t a) {
+   uint64_t x = a;
+   word h = NUMHDR, t = TNUM;
    if (a < 0) {
-      if (a < 0-FMAX)
-         exit(42);
-      return FN(0-a);
-   } else if (a > FMAX) {
-      word *r;
-      if (a >= (int64_t)1 << FBITS * 2)
-         exit(42);
-      allocate(6, r);
-      r[0] = make_header(3, 40);
-      r[4] = F(a>>FBITS);
-      r[2] = (word) (r + 3);
-      r[3] = make_header(3, 40);
-      r[1] = F(a&FMAX);
-      r[5] = INULL;
-      return (word) r;
+      h = NUMNHDR;
+      t = TNUMN;
+      x = -a;
+   }
+   if (x > FMAX) {
+      word *p, r = (word)fp;
+      do {
+         allocate(3, p);
+         p[0] = h;
+         p[1] = F(x & FMAX);
+         p[2] = (word)fp;
```

Comment:
The sign bit is only in the first header of a bignum, so that math functions can safely use the tail of any number as part of another one after handling the sign. You could for example have h = NUMHDR here to do it.